### PR TITLE
resolve: dedup entries in /etc/hosts

### PR DESCRIPTION
**Pull request by yuwata**
Fixes oss-fuzz#47708 (https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=47708).

### CODE REVIEW

**Comment by keszybz** [comment]
I don't quite understand the bug. When executed without sanitizers or valgrind, the oss-fuzz sample finishes in 0.35 s. On oss-fuzz it hits the timeout of 60 s, but I think this is just a problem with inefficiency of asan.

With valgrind it takes forever too:
`total heap usage: 1,655,134 allocs, 1,655,133 frees, 366,840,347,723 bytes allocated`

I think we want to allow bigger samples. We know that people use `/etc/hosts` to maintain denylists and such, and it's good to be able to test those.

File: src/resolve/fuzz-etc-hosts.c [hunk 1]
```diff
@@ -11,6 +11,9 @@ int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
         if (!getenv("SYSTEMD_LOG_LEVEL"))
                 log_set_max_level(LOG_CRIT);
 
+        if (size > 64*1024)
```
Comment (keszybz) [inline src/resolve/fuzz-etc-hosts.c]: `if (HAS_FEATURE_ADDRESS_SANITIZER && size > …)` ?

Comment (keszybz) [reply src/resolve/fuzz-etc-hosts.c]: I'm not *convinced* that this is the way to go. I just don't want to disable the test too much.

**Comment by mrc0mmand** [comment]
```
TEST-75-RESOLVED RUN: Tests for systemd-resolved
+ env --unset=UNIFIED_CGROUP_HIERARCHY --unset=SYSTEMD_NSPAWN_UNIFIED_HIERARCHY timeout --foreground 900 /systemd-meson-build/systemd-nspawn --register=no --kill-signal=SIGKILL --directory=/var/tmp/systemd-test-TEST-75-RESOLVED/root --setenv=SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-75.units:/usr/lib/systemd/tests/testdata/units: --machine=TEST-75 --private-network /usr/lib/systemd/systemd systemd.unit=testsuite.target systemd.wants=testsuite-75.service systemd.wants=end.service
Spawning container TEST-75 on /var/tmp/systemd-test-TEST-75-RESOLVED/root.
Press Ctrl-] three times within 1s to kill container.
/usr/lib/systemd/systemd: error while loading shared libraries: libpam.so.0: cannot open shared object file: Error 117
Container TEST-75 failed with error code 127.
E: nspawn failed with exit code 127
```

huh

**Comment by mrc0mmand** [comment]
Also:

```
[ 3893.715874] testsuite-75.sh[357]: + host 0.in-addr.arpa
[ 3893.717275] testsuite-75.sh[358]: + tee /tmp/tmp.oCaKXP3f2R
[ 3893.723605] systemd-resolved[232]: Received dns UDP packet of size 32, ifindex=0, ttl=64, fragsize=0, sender=127.0.0.1, destination=127.0.0.53
[ 3893.723741] systemd-resolved[232]: Got DNS stub UDP query packet for id 61699
[ 3893.723816] systemd-resolved[232]: Looking up RR for 0.in-addr.arpa IN A.
[ 3893.723885] systemd-resolved[232]: Failed to start query: Invalid argument
[ 3893.723954] systemd-resolved[232]: Sending response packet with id 61699 on interface 1/AF_INET of size 32.
[ 3893.724567] testsuite-75.sh[358]: Host 0.in-addr.arpa not found: 2(SERVFAIL)
```

**Comment by yuwata** [comment]
Should be fixed now. PTAL.

File: src/resolve/resolved-etc-hosts.c [hunk 2]
```diff
@@ -239,16 +245,10 @@ static void strip_localhost(EtcHosts *hosts) {
...
-                        n = hashmap_remove(hosts->by_name, *i);
-                        if (n)
-                                etc_hosts_item_by_name_free(n);
-                }
+                STRV_FOREACH(i, item->names)
+                        etc_hosts_item_by_name_free(hashmap_remove(hosts->by_name, *i));
 
-                assert_se(hashmap_remove(hosts->by_address, local_in_addrs + j) == item);
-                etc_hosts_item_by_address_free(item);
+                etc_hosts_item_by_address_free(hashmap_remove(hosts->by_address, local_in_addrs + j));
```
Comment (keszybz) [inline src/resolve/resolved-etc-hosts.c]: It doesn't matter _too_ much, but this `assert_se` would quite useful. It was also helping the reader understand the relationship between the two hashmaps.

File: src/resolve/resolved-etc-hosts.c [hunk 3]
```diff
@@ -464,3 +470,26 @@ int manager_etc_hosts_lookup(Manager *m, DnsQuestion* q, DnsAnswer **answer) {
 
         return found_a || found_aaaa;
 }
+
+int manager_etc_hosts_lookup(Manager *m, DnsQuestion *q, DnsAnswer **answer) {
+        struct in_addr_data k = {};
```
Comment (keszybz) [inline src/resolve/resolved-etc-hosts.c]: `dns_name_address()` always sets output if return value is positive, so this initialization is not necessary.

File: src/resolve/resolved-etc-hosts.c [hunk 4]
```diff
@@ -130,7 +146,7 @@ static int parse_line(EtcHosts *hosts, unsigned nr, const char *line) {
                         /* Optimize the case where we don't need to store any addresses, by storing
                          * only the name in a dedicated Set instead of the hashmap */
 
-                        r = set_ensure_consume(&hosts->no_address, &dns_name_hash_ops, TAKE_PTR(name));
+                        r = set_ensure_consume(&hosts->no_address, &dns_name_hash_ops_free, TAKE_PTR(name));
```
Comment (keszybz) [inline src/resolve/resolved-etc-hosts.c:149]: Hmm, this this chunk sneak in from another commit?

**Review by keszybz — COMMENTED** [review]
Looks nice. I was wondering whether it makes sense to introduce a more complicated data structure, even if most people have quite short `hosts` files. But I think it's reasonable: for the majority of cases, when there's a small number of entries, this doesn't make much difference. Maybe a few hundred bytes one way or another. OTOH, for the cases where there's a larger number of entries, this makes things much more robust.

**Comment by yuwata** [comment]
@keszybz Thank you for the review. All comments are addressed. Upgrading the green label.

**Comment by keszybz** [comment]
CentOS CI (Arch Linux + sanitizers) — TEST-29-PORTABLE.
jammy-ppc64el — TEST-30-ONCLOCKCHANGE:              FAIL     (1202 s)

**Pull request merged**